### ISSUE TITLE
[BACKFILL] We shouldn't be waiting for a whole batch of dag runs to be finished before triggering new ones

### ISSUE THREAD

**Issue by qcha41** [apache/airflow#38954]
### Apache Airflow version

2.9.0

### What happened?

When you launch a backfill command with the Airflow CLI, it will first trigger a batch of several dag runs in parallel (max number fixed by the parameter max_active_runs). The problem is that the backfill scheduler is waiting for all dag runs of that first batch to be finished before triggering a new batch of dag runs. 

This means that if all dag runs are not taking the same time to complete, we are unnecessarily losing processing time waiting for the slowest dag run before triggering new ones, which has a serious impact on the overall backfill operation execution time.

I don't understand why backfill jobs do not behave as when we clear mulitple tasks : as soon as a slot becomes free, the following pending task is triggered immediately.

### What you think should happen instead?

Backfill slot allocation management should be the same as the classic "clear" slot allocation management, in order to trigger new dag runs as soon as a previous one is finished, to avoid losing time unecessarily.

### How to reproduce

dag.py
```python
from datetime import datetime, timedelta
from random import randint
from time import sleep
from airflow.decorators import dag, task

# Define the default arguments
default_args = {
    'owner': 'airflow',
    'depends_on_past': False,
    'start_date': datetime(2025, 4, 11),
    'retries': 1,
    'retry_delay': timedelta(minutes=5),
}

# Define the Python function to wait for a random time
@task
def random_wait():
    wait_time = randint(1, 60)
    print(f"Waiting for {wait_time} seconds...")
    sleep(wait_time)

# Define the DAG
@dag(default_args=default_args, schedule_interval=timedelta(minutes=1), dag_id='random_wait_dag')
def random_wait_dag():
    task_1 = random_wait()
    task_2 = random_wait()
    task_3 = random_wait()

    task_1 >> task_2 >> task_3

dag = random_wait_dag()
```

Start backfill
```bash
docker compose run airflow-cli dags backfill -s 2024-04-11T00:00:00 -e 2024-04-11T01:00:00 random_wait_dag
```

Observe behaviour
![1](https://github.com/apache/airflow/assets/31767137/6370fca4-e5af-495f-b307-18fbc6b6bb0e)
![4](https://github.com/apache/airflow/assets/31767137/b2f13bb6-b2d4-4008-9e8d-8ef21b120434)
![6](https://github.com/apache/airflow/assets/31767137/7c7743b6-0174-4cc0-97dd-ed6b0c0ef95f)
![7](https://github.com/apache/airflow/assets/31767137/999c85fd-56a4-4717-b6ab-a375b1090b07)
![8](https://github.com/apache/airflow/assets/31767137/7630a98c-b2b4-4a48-a8a3-3480d32e4da0)

### Operating System

debian

### Deployment

Docker-Compose

### Are you willing to submit PR?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/apache/airflow/blob/main/CODE_OF_CONDUCT.md)

**Comment by denis-sanches** [comment]
Hi! It seems that I'm facing the exact same behaviour on version 2.10.3.

**Comment by potiuk** [comment]
Backfill is completely rewritten for Airflow 3. I guess we can close that one @dstandish as "won't fix - switch to Airflow 3 for better backfill experience".

More details in https://cwiki.apache.org/confluence/display/AIRFLOW/AIP-78+Scheduler-managed+backfill

**Comment by potiuk** [comment]
Closing this provisionally. We can always re-open if we think we can do anything else than get users to migrate to Airlfow 3 (but it's unlikely as we do not implement any features in Airflow 2 any more).

**Comment by dstandish** [comment]
@potiuk  you scared the crap out of me, i thought this was a bug with new backfill...

only a matter of time though ....

**Comment by potiuk** [comment]
> @potiuk you scared the crap out of me, i thought this was a bug with new backfill...
> 
> only a matter of time though ....

:scream:
